File: src/cpp_version/lexer.cpp
#include "lexer.h"
#include <cctype>
#include <cstdlib>

std::unordered_map<std::string, TokenKind> Lexer::keywords_ = {
    {"boolean", TokenKind::KW_BOOLEAN}, {"class", TokenKind::KW_CLASS},
    {"false", TokenKind::KW_FALSE},     {"true", TokenKind::KW_TRUE},
    {"int", TokenKind::KW_INT},         {"main", TokenKind::KW_MAIN},
    {"public", TokenKind::KW_PUBLIC},   {"static", TokenKind::KW_STATIC},
    {"String", TokenKind::KW_STRING},   {"void", TokenKind::KW_VOID},
    {"return", TokenKind::KW_RETURN},   {"if", TokenKind::KW_IF},
    {"else", TokenKind::KW_ELSE},       {"while", TokenKind::KW_WHILE},
    {"new", TokenKind::KW_NEW},         {"length", TokenKind::KW_LENGTH},
    {"private", TokenKind::KW_PRIVATE},
};

Lexer::Lexer(const std::string& source) : src_(source) {}

char Lexer::current() const {
    if (pos_ >= src_.size()) return '\0';
    return src_[pos_];
}

char Lexer::advance() {
    char c = current();
    if (c == '\n') line_++;
    pos_++;
    return c;
}

bool Lexer::match(char c) {
    if (current() == c) { advance(); return true; }
    return false;
}

void Lexer::skipWhitespaceAndComments() {
    while (pos_ < src_.size()) {
        char c = current();
        if (c == ' ' || c == '\t' || c == '\r' || c == '\n') {
            advance();
        } else if (c == '/' && pos_ + 1 < src_.size() && src_[pos_ + 1] == '/') {
            // single-line comment
            while (pos_ < src_.size() && current() != '\n') advance();
        } else if (c == '/' && pos_ + 1 < src_.size() && src_[pos_ + 1] == '*') {
            // block comment
            advance(); advance(); // skip /*
            while (pos_ < src_.size()) {
                if (current() == '*' && pos_ + 1 < src_.size() && src_[pos_ + 1] == '/') {
                    advance(); advance();
                    break;
                }
                advance();
            }
        } else {
            break;
        }
    }
}

Token Lexer::makeToken(TokenKind kind, const std::string& text) {
    Token t;
    t.kind = kind;
    t.text = text;
    t.line = line_;
    return t;
}

bool Lexer::tryMatchSequence(const std::string& seq) {
    if (pos_ + seq.size() > src_.size()) return false;
    for (size_t i = 0; i < seq.size(); i++) {
        if (src_[pos_ + i] != seq[i]) return false;
    }
    // Make sure next char after sequence isn't alphanumeric (word boundary)
    size_t end = pos_ + seq.size();
    if (end < src_.size() && (std::isalnum(src_[end]) || src_[end] == '_'))
        return false;
    for (size_t i = 0; i < seq.size(); i++) {
        if (src_[pos_] == '\n') line_++;
        pos_++;
    }
    return true;
}

Token Lexer::scanIdentifierOrKeyword() {
    size_t start = pos_;
    while (pos_ < src_.size() && (std::isalnum(current()) || current() == '_'))
        advance();
    std::string word = src_.substr(start, pos_ - start);

    // Check for compound keywords: System.out.print, System.out.println, Integer.parseInt
    if (word == "System") {
        size_t saved_pos = pos_;
        int saved_line = line_;
        // Try to match .out.println first (longer match)
        if (tryMatchSequence(".out.println"))
            return makeToken(TokenKind::SYSTEM_OUT_PRINTLN, "System.out.println");
        pos_ = saved_pos; line_ = saved_line;
        if (tryMatchSequence(".out.print"))
            return makeToken(TokenKind::SYSTEM_OUT_PRINT, "System.out.print");
        pos_ = saved_pos; line_ = saved_line;
    } else if (word == "Integer") {
        size_t saved_pos = pos_;
        int saved_line = line_;
        if (tryMatchSequence(".parseInt"))
            return makeToken(TokenKind::INTEGER_PARSEINT, "Integer.parseInt");
        pos_ = saved_pos; line_ = saved_line;
    }

    auto it = keywords_.find(word);
    if (it != keywords_.end())
        return makeToken(it->second, word);
    return makeToken(TokenKind::IDENTIFIER, word);
}

Token Lexer::scanNumber() {
    size_t start = pos_;
    while (pos_ < src_.size() && std::isdigit(current()))
        advance();
    std::string num = src_.substr(start, pos_ - start);
    Token t = makeToken(TokenKind::INTEGER_LITERAL, num);
    t.int_value = std::atoi(num.c_str());
    return t;
}

Token Lexer::scanString() {
    advance(); // skip opening "
    std::string val = "\"";
    while (pos_ < src_.size() && current() != '"') {
        if (current() == '\\') {
            val += advance(); // backslash
            if (pos_ < src_.size()) val += advance(); // escaped char
        } else {
            val += advance();
        }
    }
    if (pos_ < src_.size()) advance(); // skip closing "
    val += '"';
    return makeToken(TokenKind::STRING_LITERAL, val);
}

Token Lexer::scanRawToken() {
    skipWhitespaceAndComments();
    if (pos_ >= src_.size())
        return makeToken(TokenKind::TOK_EOF, "");

    char c = current();

    if (std::isalpha(c) || c == '_') return scanIdentifierOrKeyword();
    if (std::isdigit(c)) return scanNumber();
    if (c == '"') return scanString();

    advance();
    switch (c) {
        case '&': if (match('&')) return makeToken(TokenKind::TOK_AND, "&&");
                  return makeToken(TokenKind::TOK_ERROR, "&");
        case '|': if (match('|')) return makeToken(TokenKind::TOK_OR, "||");
                  return makeToken(TokenKind::TOK_ERROR, "|");
        case '=': if (match('=')) return makeToken(TokenKind::TOK_EQ, "==");
                  return makeToken(TokenKind::TOK_ASSIGN, "=");
        case '!': if (match('=')) return makeToken(TokenKind::TOK_NEQ, "!=");
                  return makeToken(TokenKind::TOK_NOT, "!");
        case '<': if (match('=')) return makeToken(TokenKind::TOK_LTE, "<=");
                  return makeToken(TokenKind::TOK_LT, "<");
        case '>': if (match('=')) return makeToken(TokenKind::TOK_GTE, ">=");
                  return makeToken(TokenKind::TOK_GT, ">");
        case '+': return makeToken(TokenKind::TOK_PLUS, "+");
        case '-': return makeToken(TokenKind::TOK_MINUS, "-");
        case '*': return makeToken(TokenKind::TOK_STAR, "*");
        case '/': return makeToken(TokenKind::TOK_SLASH, "/");
        case '.': return makeToken(TokenKind::TOK_DOT, ".");
        case ';': return makeToken(TokenKind::TOK_SEMI, ";");
        case '(': return makeToken(TokenKind::TOK_LPAREN, "(");
        case ')': return makeToken(TokenKind::TOK_RPAREN, ")");
        case '{': return makeToken(TokenKind::TOK_LBRACE, "{");
        case '}': return makeToken(TokenKind::TOK_RBRACE, "}");
        case ',': return makeToken(TokenKind::TOK_COMMA, ",");
        case '[': return makeToken(TokenKind::TOK_LBRACKET, "[");
        case ']': return makeToken(TokenKind::TOK_RBRACKET, "]");
        default:  return makeToken(TokenKind::TOK_ERROR, std::string(1, c));
    }
}

void Lexer::fillBuffer(int n) {
    while ((int)buffer_.size() < n) {
        buffer_.push_back(scanRawToken());
    }
}

Token Lexer::nextToken() {
    if (!buffer_.empty()) {
        Token t = buffer_.front();
        buffer_.erase(buffer_.begin());
        return t;
    }
    return scanRawToken();
}

Token Lexer::peek() {
    fillBuffer(1);
    return buffer_[0];
}

Token Lexer::peekNext() {
    fillBuffer(2);
    return buffer_[1];
}

File: src/src/lexer.rs
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenKind {
    // Keywords
    KwBoolean,
    KwClass,
    KwFalse,
    KwTrue,
    KwInt,
    KwMain,
    KwPublic,
    KwStatic,
    KwString,
    KwVoid,
    KwReturn,
    KwIf,
    KwElse,
    KwWhile,
    KwNew,
    KwLength,
    KwPrivate,
    // Compound keywords
    SystemOutPrint,
    SystemOutPrintln,
    IntegerParseInt,
    // Operators
    TokAnd,
    TokOr,
    TokEq,
    TokNeq,
    TokLte,
    TokGte,
    TokLt,
    TokGt,
    TokPlus,
    TokMinus,
    TokStar,
    TokSlash,
    TokNot,
    // Punctuation
    TokDot,
    TokSemi,
    TokLparen,
    TokRparen,
    TokLbrace,
    TokRbrace,
    TokComma,
    TokAssign,
    TokLbracket,
    TokRbracket,
    // Literals
    IntegerLiteral,
    StringLiteral,
    Identifier,
    // Special
    TokEof,
    TokError,
}

#[derive(Debug, Clone)]
pub struct Token {
    pub kind: TokenKind,
    pub text: String,
    pub line: i32,
    pub int_value: i32,
}

pub struct Lexer {
    src: Vec<u8>,
    pos: usize,
    line: i32,
    buffer: std::collections::VecDeque<Token>,
}

impl Lexer {
    pub fn new(source: &str) -> Self {
        Lexer {
            src: source.as_bytes().to_vec(),
            pos: 0,
            line: 1,
            buffer: std::collections::VecDeque::new(),
        }
    }

    fn current(&self) -> u8 {
        if self.pos >= self.src.len() {
            0
        } else {
            self.src[self.pos]
        }
    }

    fn advance(&mut self) -> u8 {
        let c = self.current();
        if c == b'\n' {
            self.line += 1;
        }
        self.pos += 1;
        c
    }

    fn match_char(&mut self, c: u8) -> bool {
        if self.current() == c {
            self.advance();
            true
        } else {
            false
        }
    }

    fn skip_whitespace_and_comments(&mut self) {
        while self.pos < self.src.len() {
            let c = self.current();
            if c == b' ' || c == b'\t' || c == b'\r' || c == b'\n' {
                self.advance();
            } else if c == b'/'
                && self.pos + 1 < self.src.len()
                && self.src[self.pos + 1] == b'/'
            {
                while self.pos < self.src.len() && self.current() != b'\n' {
                    self.advance();
                }
            } else if c == b'/'
                && self.pos + 1 < self.src.len()
                && self.src[self.pos + 1] == b'*'
            {
                self.advance();
                self.advance();
                while self.pos < self.src.len() {
                    if self.current() == b'*'
                        && self.pos + 1 < self.src.len()
                        && self.src[self.pos + 1] == b'/'
                    {
                        self.advance();
                        self.advance();
                        break;
                    }
                    self.advance();
                }
            } else {
                break;
            }
        }
    }

    fn make_token(&self, kind: TokenKind, text: &str) -> Token {
        Token {
            kind,
            text: text.to_string(),
            line: self.line,
            int_value: 0,
        }
    }

    fn try_match_sequence(&mut self, seq: &[u8]) -> bool {
        if self.pos + seq.len() > self.src.len() {
            return false;
        }
        for i in 0..seq.len() {
            if self.src[self.pos + i] != seq[i] {
                return false;
            }
        }
        let end = self.pos + seq.len();
        if end < self.src.len()
            && (self.src[end].is_ascii_alphanumeric() || self.src[end] == b'_')
        {
            return false;
        }
        for _ in 0..seq.len() {
            if self.src[self.pos] == b'\n' {
                self.line += 1;
            }
            self.pos += 1;
        }
        true
    }

    fn scan_identifier_or_keyword(&mut self) -> Token {
        let start = self.pos;
        while self.pos < self.src.len()
            && (self.src[self.pos].is_ascii_alphanumeric() || self.src[self.pos] == b'_')
        {
            self.advance();
        }
        let word = std::str::from_utf8(&self.src[start..self.pos])
            .unwrap()
            .to_string();

        if word == "System" {
            let saved_pos = self.pos;
            let saved_line = self.line;
            if self.try_match_sequence(b".out.println") {
                return self.make_token(TokenKind::SystemOutPrintln, "System.out.println");
            }
            self.pos = saved_pos;
            self.line = saved_line;
            if self.try_match_sequence(b".out.print") {
                return self.make_token(TokenKind::SystemOutPrint, "System.out.print");
            }
            self.pos = saved_pos;
            self.line = saved_line;
        } else if word == "Integer" {
            let saved_pos = self.pos;
            let saved_line = self.line;
            if self.try_match_sequence(b".parseInt") {
                return self.make_token(TokenKind::IntegerParseInt, "Integer.parseInt");
            }
            self.pos = saved_pos;
            self.line = saved_line;
        }

        let kind = match word.as_str() {
            "boolean" => TokenKind::KwBoolean,
            "class" => TokenKind::KwClass,
            "false" => TokenKind::KwFalse,
            "true" => TokenKind::KwTrue,
            "int" => TokenKind::KwInt,
            "main" => TokenKind::KwMain,
            "public" => TokenKind::KwPublic,
            "static" => TokenKind::KwStatic,
            "String" => TokenKind::KwString,
            "void" => TokenKind::KwVoid,
            "return" => TokenKind::KwReturn,
            "if" => TokenKind::KwIf,
            "else" => TokenKind::KwElse,
            "while" => TokenKind::KwWhile,
            "new" => TokenKind::KwNew,
            "length" => TokenKind::KwLength,
            "private" => TokenKind::KwPrivate,
            _ => TokenKind::Identifier,
        };
        self.make_token(kind, &word)
    }

    fn scan_number(&mut self) -> Token {
        let start = self.pos;
        while self.pos < self.src.len() && self.src[self.pos].is_ascii_digit() {
            self.advance();
        }
        let num_str = std::str::from_utf8(&self.src[start..self.pos])
            .unwrap()
            .to_string();
        let val: i32 = num_str.parse().unwrap_or(0);
        let mut t = self.make_token(TokenKind::IntegerLiteral, &num_str);
        t.int_value = val;
        t
    }

    fn scan_string(&mut self) -> Token {
        self.advance(); // skip opening "
        let mut val = String::from("\"");
        while self.pos < self.src.len() && self.current() != b'"' {
            if self.current() == b'\\' {
                val.push(self.advance() as char);
                if self.pos < self.src.len() {
                    val.push(self.advance() as char);
                }
            } else {
                val.push(self.advance() as char);
            }
        }
        if self.pos < self.src.len() {
            self.advance(); // skip closing "
        }
        val.push('"');
        self.make_token(TokenKind::StringLiteral, &val)
    }

    fn scan_raw_token(&mut self) -> Token {
        self.skip_whitespace_and_comments();
        if self.pos >= self.src.len() {
            return self.make_token(TokenKind::TokEof, "");
        }

        let c = self.current();

        if c.is_ascii_alphabetic() || c == b'_' {
            return self.scan_identifier_or_keyword();
        }
        if c.is_ascii_digit() {
            return self.scan_number();
        }
        if c == b'"' {
            return self.scan_string();
        }

        self.advance();
        match c {
            b'&' => {
                if self.match_char(b'&') {
                    self.make_token(TokenKind::TokAnd, "&&")
                } else {
                    self.make_token(TokenKind::TokError, "&")
                }
            }
            b'|' => {
                if self.match_char(b'|') {
                    self.make_token(TokenKind::TokOr, "||")
                } else {
                    self.make_token(TokenKind::TokError, "|")
                }
            }
            b'=' => {
                if self.match_char(b'=') {
                    self.make_token(TokenKind::TokEq, "==")
                } else {
                    self.make_token(TokenKind::TokAssign, "=")
                }
            }
            b'!' => {
                if self.match_char(b'=') {
                    self.make_token(TokenKind::TokNeq, "!=")
                } else {
                    self.make_token(TokenKind::TokNot, "!")
                }
            }
            b'<' => {
                if self.match_char(b'=') {
                    self.make_token(TokenKind::TokLte, "<=")
                } else {
                    self.make_token(TokenKind::TokLt, "<")
                }
            }
            b'>' => {
                if self.match_char(b'=') {
                    self.make_token(TokenKind::TokGte, ">=")
                } else {
                    self.make_token(TokenKind::TokGt, ">")
                }
            }
            b'+' => self.make_token(TokenKind::TokPlus, "+"),
            b'-' => self.make_token(TokenKind::TokMinus, "-"),
            b'*' => self.make_token(TokenKind::TokStar, "*"),
            b'/' => self.make_token(TokenKind::TokSlash, "/"),
            b'.' => self.make_token(TokenKind::TokDot, "."),
            b';' => self.make_token(TokenKind::TokSemi, ";"),
            b'(' => self.make_token(TokenKind::TokLparen, "("),
            b')' => self.make_token(TokenKind::TokRparen, ")"),
            b'{' => self.make_token(TokenKind::TokLbrace, "{"),
            b'}' => self.make_token(TokenKind::TokRbrace, "}"),
            b',' => self.make_token(TokenKind::TokComma, ","),
            b'[' => self.make_token(TokenKind::TokLbracket, "["),
            b']' => self.make_token(TokenKind::TokRbracket, "]"),
            _ => self.make_token(TokenKind::TokError, &(c as char).to_string()),
        }
    }

    fn fill_buffer(&mut self, n: usize) {
        while self.buffer.len() < n {
            let tok = self.scan_raw_token();
            self.buffer.push_back(tok);
        }
    }

    pub fn next_token(&mut self) -> Token {
        if let Some(t) = self.buffer.pop_front() {
            t
        } else {
            self.scan_raw_token()
        }
    }

    pub fn peek(&mut self) -> Token {
        self.fill_buffer(1);
        self.buffer[0].clone()
    }

    pub fn peek_next(&mut self) -> Token {
        self.fill_buffer(2);
        self.buffer[1].clone()
    }
}
